src/lib/provable/bigint.ts

```diff
@@ -0,0 +1,897 @@
+
+          // Loop until newR is equal to zero
+          while (newR !== 0n) {
+            const quotient = r / newR;
+
+            [t, newT] = [newT, t - quotient * newT];
+            [r, newR] = [newR, r - quotient * newR];
+          }
+
+          // If r is bigger than 1, a is not invertible
+          if (r > 1n) {
+            throw new Error('a is not invertible');
+          }
+
+          // If t is smaller than zero, add modulus
+          if (t < 0n) {
+            t = t + p;
+          }
+
+          return { res: ProvableBigInt_.fromBigInt(t) };
+        }
+      );
+
+      res.mul(this).assertEquals(ProvableBigInt_.one());
+
+      return res;
+    }
+
+    /**
+     * Computes the additive inverse of a ProvableBigInt
+     * Cost: Cheap
+     * @returns The additive inverse as a ProvableBigInt
+     */
+    negate(): ProvableBigInt_ {
+      let { negation } = Provable.witness(provable({ negation: ProvableBigInt_ }), () => {
+        let thisVal = this.toBigInt();
+        let p = this.Constructor.modulus.toBigInt();
+        let negVal = 0n;
+        if (thisVal !== 0n) {
+          negVal = p - thisVal;
+        }
+        return {
+          negation: ProvableBigInt_.fromBigInt(negVal),
+        };
+      });
+
+      this.add(negation).assertEquals(ProvableBigInt_.zero());
+
+      return negation;
+    }
+
+    /**
+     * Computes the power of a ProvableBigInt raised to an exponent
+     * Cost: Expensive
+     * @param exp The exponent
+     * @returns The result as a ProvableBigInt
+     */
+    pow(exp: ProvableBigInt_): ProvableBigInt_ {
+      const exponentBits = exp.toBits();
+      const processChunk = function* (bits: Bool[], chunkSize: number) {
+        for (let i = 0; i < bits.length; i += chunkSize) {
+          yield bits.slice(i, i + chunkSize);
+        }
+      };
+
+      let result = ProvableBigInt_.one();
+      let base = this.clone();
+
+      for (const chunk of processChunk(exponentBits, 100)) {
+        for (const bit of chunk) {
+          result = Provable.if(bit, ProvableBigInt_, result.mul(base), result);
+          base = base.mul(base);
+        }
+      }
+
+      return result;
+    }
+
+    /**
+     * Computes the square root of a Provable BigInt
+     * Cost: Cheap
+     * @returns The square root as a ProvableBigInt
+     */
+    sqrt(): ProvableBigInt_ {
+      let r = Provable.witness(ProvableBigInt_, () => {
+        const p = this.Constructor.modulus.toBigInt();
+        // Special cases
+        // Case 1: If input is 0, square root is 0
+        if (this.toBigInt() === 0n) return ProvableBigInt_.fromBigInt(0n);
+
+        // Case 2: If p ≡ 3 (mod 4), we can use a simpler formula
+        // In this case, the square root is a^((p+1)/4) mod p
+        if (p % 4n === 3n) {
+          const pplusonedivfour = (p + 1n) / 4n;
+          const pplusonedivfour_reduced = pplusonedivfour % p;
+          const sqrt = modularExponentiation(this.toBigInt(), pplusonedivfour_reduced, p);
```

Comment:
Nit: rename following camelCase naming conventions.
```suggestion
          const sqrt = modularExponentiation(this.toBigInt(), pPlusOneDivFourReduced, p);
```